text/0000-combined-restorer-analyzer-phases.md

```diff
@@ -0,0 +1,130 @@
+# Meta
+[meta]: #meta
+- Name: Combined restorer/analyzer and cacher/exporter phases.
+- Start Date: 2019-08-02
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+Though not formally part of the spec, the project description and the
+lifecycle reference implementation contain the following four phases:
+
+- `/lifecycle/restorer` (considered an optional platform step)
+- `/lifecycle/analyzer`
+- `/lifecycle/exporter`
+- `/lifecycle/cacher` (considered an optional platform step)
+
+This proposal seeks to combine `restorer` with `analyzer` (to be named
+`restorer`) and `cacher` with `exporter` (to be named `exporter`).
+
+# Motivation
+[motivation]: #motivation
+
+In the current lifecycle reference implementation, the `restorer` phase
+fetches cached layers from a previous cache image. These layers can be
+arbitrarily large and, when using a remote repository, can represent a large
+amount of overhead. The `analyzer` phase runs after `restorer` and it
+may deterine that particular cache layers are no longer valid and remove them
+from the rest of the process - however, we've already paid a large cost to
+fetch them in the first place.
+
+The determination of cache layer validity can frequently be performed
+through the analysis of metadata: comparing metadata from the cache image with
+metadata from the previous application image. However, this comparison can
+only be performed with a combined `restorer` and `analyzer` phase.
+
+A parallel optimization potential exists when combining the `cacher` and
+`exporter` phases.
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Combined `restorer` phase.
+
+The spec will formalize the (combined) `restorer` phase like this:
+
+```
+/lifecycle/restorer \
+  -group=[(required) group.toml file] \
+  -layers=[(required) layers directory] \
+  -image=[(optional) repository url for existing application image] \
```

Comment:
Is this equivalent to the positional argument, dubbed `repoName`, in `/lifecycle/analyzer`?

Reply:
I was wondering why this was optional, but I think maybe you're not trying to be precise here (which is fine).

Reply:
Yes, this was the item named `repoName` in the lifecycle reference implementation.

I wasn't certain about the reasoning behind choosing a positional arg over a flag - do you know?

I was trying to be accurate with required/optional, so please call these out if they are incorrect.

I set it optional because I can imagine a case where it is not possible to reference a previous application image (but you still might be restoring cache layers). Thoughts?

Reply:
That reasoning makes sense to me. Right now we will correctly handle an image tag reference that doesn't actually point at a previous image. I think it similarly makes sense to handle an omitted `-image` flag, although the cases where I would actually want to do that aren't as obvious to me.

Thinking about the inverse case, I wonder if we should also specify the `skip-layers` flag in the case that a user wants to `analyze` in order to restore [store.toml](https://github.com/buildpack/spec/blob/f08181f3e8e81e76c9d3758c0ed58303805ea8e0/buildpack.md#store.toml-toml) (when we support it) or take advantage of performance efficiencies during export, but doesn't want to restore cached layers or layer metadata (e.g `pack build --clear-cache`). I know we don't always specify all the flags but this one seems important to require, so the spec can be fully implemented.